// protocols/tfa.js
"use strict";

/*
Copyright (c) 2017-2020 Ramón Baas

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/*
  TFA 30.3139 Protocol
*/

const utils = require('utils');

class TFA extends utils.WeatherSignal {

	constructor() {
    super({
			id: 'tfa',
			name: 'TFA',
			signal: 'TFA',
			hint: {
				en: 'Support for TFA 30.3139',
				nl: 'Ondersteuning voor TFA 30.3139',
			}
		})
	}

	// Parse the data according to the TFA protocol
	parse(payLoad) {
		let result;
		let str = payLoad.join('');
		let sum = -1
		for (let i = 4; i < 28; i += 4) {
			sum += Number(utils.bin2dec(str.slice(i, i + 4)));
		}
		let crc = Number(utils.bin2dec(str.slice(0, 4)));
		if ((sum % 16) === crc) {
			result = this.decodeTFA(str);
		} else {
			result = new Error('CRC mismatch (got ' + sum % 16 + ' expected ' + crc + ')');
		}

		return result;
	}

	// Decode the data
	/*
		a a a a  i i i i i i i i  t t t t t t t t  t t t t d d e e
		0 1 2 3  4 5 6 7 8 9 0 1 12 3 4 5 6 7 8 9 20 1 2 3 4 5 6 7
	*/
	decodeTFA(data) {
		let id = utils.bin2dec(data.slice(4, 12)); // needs to be string for device setting
		let channel = Number(utils.bin2dec(data.slice(24, 26)));
		let temp = utils.bin2dec(data.slice(12, 24));
		temp = (temp > 2048 ? temp - 4096 : temp) / 10
		let result = {
			id: id,
			channel: channel,
			data: { temperature: Number(temp), lowbattery: data[26] === '0' }
		};

		return result;
	}

}


module.exports = {
	tfa: TFA
};
