// protocols/upm.js
"use strict";

/*
Copyright (c) 2017-2020 Ramón Baas

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/*
  UPM/ESIC Protocol

  Uses Differential Manchester encoding (Biphase Mark Code)

  See http://ala-paavola.fi/jaakko/doku.php?id=wt450h
  and http://wiki.nethome.nu/doku.php/upmprotocol
*/

const utils = require('utils');


class UPM extends utils.WeatherSignal {

	constructor() {
    super({
      id: 'upm',
			name: 'UPM/Esic',
      signal: 'UPM',
			hint: {
				en: 'Support for WT440H (UPM, Esic, DMV, Clas Ohlson)',
				nl: 'Ondersteuning voor WT440H (UPM, Esic, DMV, Clas Ohlson)',
			}
    })
  }

  // Parse the data according to the UPM/Esic protocol
  parse(payLoad) {
  	// 0:	[2000, 1000], 1: [1000, 1000], 2: [2000, 2000], 3: [1000, 2000]
  	const map = [ '0', '1', '00', '10' ];
  	let result;
  	// Transform to regular bitstream
  	let decVal = '';
  	for (let i = 0; i < payLoad.length; i++) {
  		decVal += map[payLoad[i]];
  	}
  	let len = decVal.length;
  	this.debug('New bitstream', decVal, len);
  	if (len === 32 && decVal.slice(6, 8) === '11') {
  		if (this.calcParity(decVal)) {
  			result = this.decode(decVal);
  		} else {
  			result = new Error('Parity error for UPM');
  		}
  	} else {
      result = new Error('Not a valid UPM message');
    }

  	return result;
  }

  // Calculate parity
  calcParity(values) {
  	let sum = 0;
  	for (let i = 0; i < values.length; i++) {
  		sum += Number(values[i]);
  	}
  	return (sum % 1 === 0);
  }

  // Decode the Esic data
  /*
  	c c c c  d d 1 1 b S S S  s s s s 0 P P P  p p p p p p p p  z z C C
  	0 1 2 3  4 5 6 7 8 9 0 1 12 3 4 5 6 7 8 9 20 1 2 3 4 5 6 7 28 9 9 1
  */
  decode(data) {
  	let hc = utils.bin2dec(data.slice(0, 4)); // needs to be string for device setting
  	let channel = Number(utils.bin2dec(data.slice(4, 6))) + 1;
  	let type = (hc === '10' && channel === 2) ? 'W' : (hc === '10' && channel === 3) ? 'R' : 'TH';
  	let result = {
  		id: hc,
  		channel: channel
  	}
  	if (type === 'TH') {
  		let temp = (utils.bin2dec(data.slice(17, 28)) / 16 - 50).toString();
  		let dot = temp.indexOf('.');
  		result.data = {
  			temperature: Number((dot >= 0 ? temp.slice(0, dot + 2) : temp)),
  			humidity: Number(utils.bin2dec(data.slice(9, 16)))
  		}
  	} else if (type === 'R') {
  		result.data = {
  			rainrate: utils.bin2dec(data.slice(17, 28)) * 0.7
  		}
  	} else if (type === 'W') {
  		result.data = {
  			direction: utils.bin2dec(data.slice(9, 16)) * 45,
  			currentspeed: utils.bin2dec(data.slice(17, 28)) / 3.6
  		}
  	}
  	result.data.lowbattery = (data[8] === '1');

  	return result;
  }

}


module.exports = {
  upm: UPM
}
