// labs/index.js
"use strict"

/*
Copyright (c) 2018 Ramón Baas

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/*
  Protocol for Labs BL999

  Based on information from https://github.com/sprilukin/lib_BL999
*/

const utils = require('utils')

var labs;

// Calculate checksum
function calcSimpleSum(data) {
	let sum = 0
	for (let i = 0; i < (data.length >> 2) - 1; i++) {
		sum += Number(utils.bin2dec(data.slice(i * 4, (i + 1) * 4).reverse().join('')))
	}
	return sum & 0xf
}

// Decode the payload
function decodeLabsData(data) {
  let id = utils.bin2dec(data.slice(0, 4).reverse().join('') + data.slice(6, 8).reverse().join(''))

  let temperature = data.slice(12, 24).reverse().join('')
  if (temperature[0] === '1') {
    // make it 32 bit, convert 2's complement
    temperature = ~~parseInt('11111111111111111111' + temperature, 2)
  } else {
    temperature = utils.bin2dec(temperature)
  }
  let humidity = 100 - (255 - utils.bin2dec(data.slice(24, 32).reverse().join('')))

	let result = {
    name: 'BL999',
		id: id,
    channel: Number(utils.bin2dec(data.slice(4, 6).join(''))),
		data: {
      temperature: Number((temperature / 10).toFixed(1)),
      humidity: humidity,
      lowbattery: data[8] === '1'
    }
	}
  labs.debug(result.data)

	return result
}

var parseLabsData = function(payLoad) {
  let result = 'Not a valid Labs message'
  let data = Array.prototype.slice.call(payLoad, 0)
  let crc = calcSimpleSum(data)
  let check = Number(utils.bin2dec(data.slice(32, 36).reverse().join('')))
  if (crc === check) {
      result = decodeLabsData(data)
  } else {
      result = 'CRC error for Labs data (' + crc + ' vs ' + check + ')'
  }
  return result
}

module.exports = {
    init: () => labs = new utils.WeatherSignal({
        id: 'labs',
        name: 'Labs',
        hint: {
            en: 'Supports the Labs BL999 sensor',
            nl: 'Ondersteunt de Labs BL999 sensor'
        },
        signal: 'Labs',
        parser: parseLabsData
    })
}
